agent/connect/ca/aws_pca.go

```diff
@@ -0,0 +1,438 @@
+			CertificateAuthorityArn: aws.String(pca.arn),
+			CertificateArn:          issueOutput.CertificateArn,
+		}
+		certOutput, err := pca.client.GetCertificate(&certInput)
+		if err != nil {
+			if err.(awserr.Error).Code() != acmpca.ErrCodeRequestInProgressException {
+				pca.logger.Printf("[ERR] error retrieving new certificate from %s: %s",
+					*issueOutput.CertificateArn, err.Error())
+				return "", fmt.Errorf("error retrieving certificate from PCA: %s", err)
+			}
+		}
+
+		if certOutput.Certificate != nil {
+			pca.logger.Printf("[DEBUG] certificate is ready, ARN is %s", *issueOutput.CertificateArn)
+
+			newCert, err := connect.ParseCert(*certOutput.Certificate)
+			if err == nil {
+				pca.logger.Printf("[DEBUG] certificate created: commonName=%s subjectKey=%s authorityKey=%s",
+					newCert.Subject.CommonName,
+					connect.HexString(newCert.SubjectKeyId),
+					connect.HexString(newCert.AuthorityKeyId))
+			}
+
+			return *certOutput.Certificate, nil
+		}
+
+		pca.logger.Printf("[DEBUG] sleeping for %s until certificate is ready", pca.sleepTime)
+		time.Sleep(pca.sleepTime)
+	}
+}
+
+func (pca *AmazonPCA) SignLeaf(csrPEM string, validity time.Duration) (string, error) {
+	certPEM, err := pca.Sign(csrPEM, LeafTemplateARN, validity)
+	return certPEM, err
+}
+
+func (pca *AmazonPCA) SignIntermediate(csrPEM string) (string, error) {
+	certPEM, err := pca.Sign(csrPEM, IntermediateTemplateARN, IntermediateValidity)
+	return certPEM, err
+}
+
+func (pca *AmazonPCA) SignRoot(csrPEM string) (string, error) {
+	certPEM, err := pca.Sign(csrPEM, RootTemplateARN, RootValidity)
+	return certPEM, err
+}
+
+func (pca *AmazonPCA) Disable() error {
+	input := acmpca.UpdateCertificateAuthorityInput{
+		CertificateAuthorityArn: aws.String(pca.arn),
+		Status:                  aws.String(acmpca.CertificateAuthorityStatusDisabled),
+	}
+	pca.logger.Printf("[INFO] disabling PCA %s", pca.arn)
+	_, err := pca.client.UpdateCertificateAuthority(&input)
+
+	return err
+}
+
+func (pca *AmazonPCA) Enable() error {
+	input := acmpca.UpdateCertificateAuthorityInput{
+		CertificateAuthorityArn: aws.String(pca.arn),
+		Status:                  aws.String(acmpca.CertificateAuthorityStatusActive),
+	}
+	pca.logger.Printf("[INFO] enabling PCA %s", pca.arn)
+	_, err := pca.client.UpdateCertificateAuthority(&input)
+
+	return err
+}
+
+func (pca *AmazonPCA) Delete() error {
+	input := acmpca.DeleteCertificateAuthorityInput{
+		CertificateAuthorityArn: aws.String(pca.arn),
+	}
+	pca.logger.Printf("[INFO] deleting PCA %s", pca.arn)
+	_, err := pca.client.DeleteCertificateAuthority(&input)
+
+	return err
+}
+
+func (pca *AmazonPCA) Undelete() error {
+	input := acmpca.RestoreCertificateAuthorityInput{
+		CertificateAuthorityArn: aws.String(pca.arn),
+	}
+	pca.logger.Printf("[INFO] undeleting PCA %s", pca.arn)
+	_, err := pca.client.RestoreCertificateAuthority(&input)
+
+	return err
+}
+
+// utility functions
+
+func GetCommonName(pcaType string, clusterId string) string {
```

Comment:
nit: do these need to be exported? Generally helpers only used inside the same package are better left unexported as it's technically a breaking change to the package interface if you ever want to change them this way.

Reply:
I actually thought they might be needed elsewhere for e.g. comparing names in different certs.